R/cellbrowser.R

```diff
@@ -0,0 +1,499 @@
+    emb <- dr[[embedding]]
+    if (is.null(x = emb)) {
+        message("Embedding ", embedding, " does not exist in Seurat object. Skipping. ")
+        next
+    }
+    #df <- slot(emb, "cell.embeddings")
+    df <-  emb@cell.embeddings
+    if (ncol(x = df) > 2) {
+      warning('Embedding ', embedding, ' has more than 2 coordinates, taking only the first 2')
+      df <- df[, 1:2]
+    }
+    colnames(x = df) <- c("x", "y")
+    df <- data.frame(cellId = rownames(x = df), df, check.names = FALSE)
+    fname <- file.path(
+      dir,
+      sprintf("%s.coords.tsv", embedding)
+    )
+    message("Writing embeddings to ", fname)
+    write.table(df[cellOrder, ], sep = "\t", file = fname, quote = FALSE, row.names = FALSE)
+    embedNames <- append(x = embedNames, values = embedding)
+  }
+  # by default, the embeddings are sorted in the object by order of creation (pca, tsne, umap).
+  # But that is usually the opposite of what users want, they want the last embedding to appear first
+  # in the UI, so reverse the order here
+  embedNames <- sort(embedNames, decreasing = TRUE)
+  embeddings.conf <- c()
+  for (embedName in embedNames) {
+      conf <- sprintf(
+        '{"file": "%s.coords.tsv", "shortLabel": "Seurat %1$s"}',
+        embedName
+      )
+      embeddings.conf <- c(embeddings.conf, conf)
+   }
+  # Export metadata
+  df <- data.frame(row.names = cellOrder, check.names = FALSE)
+  for (field in meta.fields) {
+    if (field == ".ident") {
+      df$Cluster <- idents
+      enum.fields <- c(enum.fields, "Cluster")
+    } else {
+      name <- meta.fields.names[[field]]
+      if (is.null(name)) {
+        name <- field
+      }
+      df[[name]] <- meta[[field]]
+      if (!is.numeric(df[[name]])) {
+        enum.fields <- c(enum.fields, name)
+      }
+    }
+  }
+  df <- data.frame(Cell = rownames(df), df, check.names = FALSE)
+  fname <- file.path(dir, "meta.tsv")
+  message("Writing meta data to ", fname)
+  write.table(as.matrix(df[cellOrder, ]), sep = "\t", file = fname, quote = FALSE, row.names = FALSE)
+  # Export markers
+  markers.string <- ''
+  if (is.null(markers.file)) {
+    ext <- "tsv"
+  } else {
+    ext <- tools::file_ext(markers.file)
+  }
+  file <- paste0("markers.", ext)
+  fname <- file.path(dir, file)
+  if (!is.null(markers.file) && !skip.markers) {
+    message("Copying ", markers.file, " to ", fname)
+    file.copy(markers.file, fname)
+  }
+  if (is.null(markers.file) && skip.markers) {
+    file <- NULL
+  }
+  if (is.null(markers.file) && !skip.markers) {
+    if (length(levels(idents)) > 1) {
+      markers.helper <- function(x) {
+        partition <- markers[x,]
+        ord <- order(partition$p_val_adj < 0.05, -partition$avg_logFC)
+        res <- x[ord]
+        naCount <- max(0, length(x) - markers.n)
+        res <- c(res[1:markers.n], rep(NA, naCount))
+        return(res)
+      }
+      if (.hasSlot(object, "misc") && !is.null(x = object@misc["markers"][[1]])) {
```

Comment:
See the `Misc` function from Seurat

Reply:
Does that work with Seurat2? this line will work with Seurat2.

Reply:
Let me know if you think I should use the Misc() function instead of object@misc, but object@misc seems good to me. I'll resolve the question now. Feel free to re-open.